components/style/stylist.rs

```diff
@@ -1524,86 +1465,108 @@ impl SelectorMapEntry for RevalidationSelectorAndHashes {
-/// (unfortunately, though).
-///
-/// With cousin sharing, we not only need to care about selectors in stuff like
-/// foo:first-child, but also about selectors like p:first-child foo, since the
-/// two parents may have shared style, and in that case we can test cousins
-/// whose matching depends on the selector up in the chain.
-///
-/// TODO(emilio): We can optimize when matching only siblings to only match the
-/// rightmost selector until a descendant combinator is found, I guess, and in
-/// general when we're sharing at depth `n`, to the `n + 1` sequences of
-/// descendant combinators.
-///
-/// I don't think that in presence of the bloom filter it's worth it, though.
-struct RevalidationVisitor;
+struct StylistSelectorVisitor<'a> {
+    needs_revalidation: bool,
+    passed_rightmost_selector: bool,
+    mapped_ids: &'a mut BloomFilter,
+    attribute_dependencies: &'a mut BloomFilter,
+    style_attribute_dependency: &'a mut bool,
+    state_dependencies: &'a mut ElementState,
+}
 
-impl SelectorVisitor for RevalidationVisitor {
-    type Impl = SelectorImpl;
+fn component_needs_revalidation(c: &Component<SelectorImpl>) -> bool {
+    match *c {
+        Component::AttributeInNoNamespaceExists { .. } |
+        Component::AttributeInNoNamespace { .. } |
+        Component::AttributeOther(_) |
+        Component::Empty |
+        // FIXME(bz) We really only want to do this for some cases of id
+        // selectors.  See
+        // https://bugzilla.mozilla.org/show_bug.cgi?id=1369611
+        Component::ID(_) |
+        Component::FirstChild |
+        Component::LastChild |
+        Component::OnlyChild |
+        Component::NthChild(..) |
+        Component::NthLastChild(..) |
+        Component::NthOfType(..) |
+        Component::NthLastOfType(..) |
+        Component::FirstOfType |
+        Component::LastOfType |
+        Component::OnlyOfType => {
+            true
+        },
+        Component::NonTSPseudoClass(ref p) => {
+            p.needs_cache_revalidation()
+        },
+        _ => {
+            false
+        }
+    }
+}
 
+impl<'a> SelectorVisitor for StylistSelectorVisitor<'a> {
+    type Impl = SelectorImpl;
 
-    fn visit_complex_selector(&mut self,
-                              _: SelectorIter<SelectorImpl>,
-                              combinator: Option<Combinator>) -> bool {
-        let is_sibling_combinator =
-            combinator.map_or(false, |c| c.is_sibling());
+    fn visit_complex_selector(
+        &mut self,
+        _: SelectorIter<SelectorImpl>,
+        combinator: Option<Combinator>
+    ) -> bool {
+        self.needs_revalidation =
+            self.needs_revalidation || combinator.map_or(false, |c| c.is_sibling());
+
+        // NOTE(emilio): This works properly right now because we can't store
+        // complex selectors in nested selectors, otherwise we may need to
+        // rethink this.
+        self.passed_rightmost_selector =
```

Comment:
I think the comment about how this is called before we visit any simple selectors in the non-rightmost ComplexSelector should somehow get put in here.  That's pretty important for the way we use this member.  I know you reference this again below in the comments on Component::ID.